controller/system_rollout_controller.go

```diff
@@ -0,0 +1,1843 @@
+
+	ds *datastore.DataStore
+
+	backupTargetClient     engineapi.BackupTargetClientSystemBackupInterface
+	backupTargetURL        string
+	backupTargetCredential map[string]string
+
+	systemRestore        *longhorn.SystemRestore
+	systemRestoreName    string
+	systemRestoreVersion string
+
+	downloadPath string
+	engineImage  string
+
+	extractedResources
+
+	cacheErrors util.MultiError
+	cacheSyncs  []cache.InformerSynced
+}
+
+func NewSystemRolloutController(
+	systemRestoreName string,
+	logger logrus.FieldLogger,
+	controllerID string,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	stopCh chan struct{},
+	kubeClient clientset.Interface,
+	extensionsClient apiextensionsclientset.Interface,
+) *SystemRolloutController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	c := &SystemRolloutController{
+		baseController: newBaseControllerWithQueue(SystemRolloutControllerName, logger,
+			workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+				workqueue.NewItemExponentialFailureRateLimiter(100*time.Millisecond, 2*time.Second),
+				&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(100), 1000)},
+			), SystemRolloutControllerName),
+		),
+		controllerID: controllerID,
+		stopCh:       stopCh,
+
+		kubeClient: kubeClient,
+
+		ds:            ds,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, corev1.EventSource{Component: SystemRolloutControllerName + "-controller"}),
+
+		systemRestoreName: systemRestoreName,
+	}
+
+	ds.CSIDriverInformer.AddEventHandler(c.controlleeHandler())
+	ds.DaemonSetInformer.AddEventHandler(c.namespacedControlleeHandler())
+	ds.DeploymentInformer.AddEventHandler(c.namespacedControlleeHandler())
+	cacheSyncs := []cache.InformerSynced{
+		ds.CSIDriverInformer.HasSynced,
+		ds.DaemonSetInformer.HasSynced,
+		ds.DeploymentInformer.HasSynced,
+	}
+
+	if _, err := extensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), CRDVolumeName, metav1.GetOptions{}); err == nil {
+		ds.VolumeInformer.AddEventHandler(c.controlleeHandler())
+		cacheSyncs = append(cacheSyncs, ds.VolumeInformer.HasSynced)
+	}
+	if _, err := extensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), CRDEngineImageName, metav1.GetOptions{}); err == nil {
+		ds.EngineImageInformer.AddEventHandler(c.controlleeHandler())
+		cacheSyncs = append(cacheSyncs, ds.EngineImageInformer.HasSynced)
+	}
+	if _, err := extensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(context.TODO(), CRDRecurringJobName, metav1.GetOptions{}); err == nil {
+		ds.RecurringJobInformer.AddEventHandler(c.controlleeHandler())
+		cacheSyncs = append(cacheSyncs, ds.RecurringJobInformer.HasSynced)
+	}
```

Comment:
I know you learn from the uninstall controller, but why we need these here? or just want to have a trigger event to start the system rollout?